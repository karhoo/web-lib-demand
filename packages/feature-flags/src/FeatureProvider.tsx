import React, { Component, createContext, ReactNode } from 'react'

import { Features } from './types'

type context = {
  isEnabled?(name: string): boolean
}

export const FeatureContext = createContext<context>({})

declare global {
  interface Window {
    Feature: {
      disableFeature(name: string): void
      enableFeature(name: string): void
      isEnabled(name: string): void
    }
  }
}

type Props = {
  children: ReactNode
  feature: {
    disable(name: string): void
    enable(name: string): void
    isEnabled(name: string): boolean
    init: () => Promise<void>
    listen(onChange: Function): void
  }
  spinner?: ReactNode
  fallback?: ReactNode
  logger?: {
    error(error: Error): void
  }
}

type State = {
  featureConfig: Features
  isInitialized: boolean
  requestIsFailed?: boolean
}

export class FeatureProvider extends Component<Props, State> {
  static defaultProps = {
    spinner: null,
    fallback: null,
  }

  constructor(props: Props) {
    super(props)

    this.state = {
      featureConfig: {},
      isInitialized: false,
    }

    window.Feature = {
      disableFeature: this.disableFeature,
      enableFeature: this.enableFeature,
      isEnabled: this.isEnabled,
    }
  }

  componentDidMount() {
    this.init()
  }

  public init = () => {
    const { feature, logger } = this.props

    feature.listen(this.onFeaturesChange)
    feature
      .init()
      .then(() => this.setState({ isInitialized: true }))
      .catch(error => {
        logger && logger.error(new Error(`Feature flags init error: ${error ? error.message : ''}`))

        this.setState({ requestIsFailed: true })
      })
  }

  private onFeaturesChange = (featureConfig: Features) =>
    this.setState({
      featureConfig,
    })

  public disableFeature = (name: string) => this.props.feature.disable(name)

  public enableFeature = (name: string) => this.props.feature.enable(name)

  public isEnabled = (name: string) => this.props.feature.isEnabled(name)

  public render() {
    const provider = (
      <FeatureContext.Provider value={{ isEnabled: this.isEnabled }}>
        {this.props.children}
      </FeatureContext.Provider>
    )
    const { isInitialized, requestIsFailed } = this.state
    const { spinner, fallback } = this.props

    if (isInitialized) return provider

    // in case of failed request and empty fallback all features will be either disabled
    // or enabled (according to `defaultValueForMissingKeys` option. False by default)
    if (requestIsFailed) return fallback ? fallback : provider

    return spinner
  }
}

type FeatureProps = {
  name: string
  children: ReactNode
  disabledElement: object
  logger?: {
    error: Error
  }
}

const FeatureDefaultProps = {
  children: null,
  disabledElement: null,
}

export const Feature = ({ name, children, disabledElement }: FeatureProps) => {
  return (
    <FeatureContext.Consumer>
      {({ isEnabled = () => false }) => (isEnabled(name) ? children : disabledElement)}
    </FeatureContext.Consumer>
  )
}

Feature.defaultProps = FeatureDefaultProps
