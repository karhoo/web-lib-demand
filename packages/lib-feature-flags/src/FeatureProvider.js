// eslint-disable-next-line no-unused-vars
import React, { Component, createContext } from 'react'
import PropTypes from 'prop-types'

export const FeatureContext = createContext({})

export class FeatureProvider extends Component {
  static propTypes = {
    children: PropTypes.node,
    feature: PropTypes.shape({
      disable: PropTypes.func.isRequired,
      enable: PropTypes.func.isRequired,
      init: PropTypes.func.isRequired,
    }),
    spinner: PropTypes.node,
    fallback: PropTypes.node,
  }

  static defaultProps = {
    spinner: null,
    fallback: null,
  }

  constructor(props) {
    super(props)

    this.state = {
      featureConfig: {},
      isInitialized: false,
    }

    window.Feature = {
      disableFeature: this.disableFeature,
      enableFeature: this.enableFeature,
      isEnabled: this.isEnabled,
    }
  }

  componentDidMount() {
    this.init()
  }

  init = () => {
    const { feature, logger } = this.props

    feature.listen(this.onFeaturesChange)
    feature
      .init()
      .then(() => this.setState({ isInitialized: true }))
      .catch(error => {
        logger && logger.error(new Error(`Feature flags init error: ${error ? error.message : ''}`))

        this.setState({ requestIsFailed: true })
      })
  }

  onFeaturesChange = featureConfig =>
    this.setState({
      featureConfig,
    })

  disableFeature = name => this.props.feature.disable(name)

  enableFeature = name => this.props.feature.enable(name)

  isEnabled = name => this.props.feature.isEnabled(name)

  render() {
    const provider = (
      <FeatureContext.Provider value={{ isEnabled: this.isEnabled }}>
        {this.props.children}
      </FeatureContext.Provider>
    )
    const { isInitialized, requestIsFailed } = this.state
    const { spinner, fallback } = this.props

    if (isInitialized) return provider

    // in case of failed request and empty fallback all features will be either disabled
    // or enabled (according to `defaultValueForMissingKeys` option. False by default)
    if (requestIsFailed) return fallback ? fallback : provider

    return spinner
  }
}

export const Feature = ({ name, children, disabledElement }) => {
  return (
    <FeatureContext.Consumer>
      {({ isEnabled }) => (isEnabled(name) ? children : disabledElement)}
    </FeatureContext.Consumer>
  )
}

Feature.propTypes = {
  children: PropTypes.node,
  name: PropTypes.string.isRequired,
  disabledElement: PropTypes.object,
  logger: PropTypes.shape({
    error: PropTypes.func.isRequired,
  }),
}

Feature.defaultProps = {
  children: null,
  disabledElement: null,
}
