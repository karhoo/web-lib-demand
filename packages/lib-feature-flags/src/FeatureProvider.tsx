import React, { Component, createContext, ReactNode } from 'react'

import { Features } from './types'
import { FeatureFlags } from './FeatureFlags'

type context = {
  isEnabled?(name: string): boolean
}

export const FeatureContext = createContext<context>({})

declare global {
  interface Window {
    Feature: {
      disableFeature(name: string): void
      enableFeature(name: string): void
      isEnabled(name: string): void
    }
  }
}

type Props = {
  children: ReactNode
  featureFlags: FeatureFlags
  spinner?: ReactNode
  fallback?: ReactNode
  logger?: {
    error(error: Error): void
  }
}

type State = {
  featureConfig: Features
  isInitialized: boolean
  requestIsFailed?: boolean
}

export class FeatureProvider extends Component<Props, State> {
  static defaultProps = {
    spinner: null,
    fallback: null,
  }

  constructor(props: Props) {
    super(props)

    this.state = {
      featureConfig: {},
      isInitialized: false,
    }

    window.Feature = {
      disableFeature: this.disableFeature,
      enableFeature: this.enableFeature,
      isEnabled: this.isEnabled,
    }
  }

  componentDidMount() {
    this.init()
  }

  public init = () => {
    const { featureFlags, logger } = this.props

    featureFlags.listen(this.onFeaturesChange)
    featureFlags
      .init()
      .then(() => this.setState({ isInitialized: true }))
      .catch(error => {
        logger && logger.error(new Error(`Feature flags init error: ${error ? error.message : ''}`))

        this.setState({ requestIsFailed: true })
      })
  }

  private onFeaturesChange = (featureConfig: Features) =>
    this.setState({
      featureConfig,
    })

  public disableFeature = (name: string) => this.props.featureFlags.disable(name)

  public enableFeature = (name: string) => this.props.featureFlags.enable(name)

  public isEnabled = (name: string) => this.props.featureFlags.isEnabled(name)

  public render() {
    const provider = (
      <FeatureContext.Provider value={{ isEnabled: this.isEnabled }}>
        {this.props.children}
      </FeatureContext.Provider>
    )
    const { isInitialized, requestIsFailed } = this.state
    const { spinner, fallback } = this.props

    if (isInitialized) return provider

    // in case of failed request and empty fallback all features will be either disabled
    // or enabled (according to `defaultValueForMissingKeys` option. False by default)
    if (requestIsFailed) return fallback ? fallback : provider

    return spinner
  }
}

type FeatureProps = {
  name: string
  children: ReactNode
  disabledElement: object
  logger?: {
    error: Error
  }
}

const FeatureDefaultProps = {
  children: null,
  disabledElement: null,
}

export const Feature = ({ name, children, disabledElement }: FeatureProps) => {
  return (
    <FeatureContext.Consumer>
      {({ isEnabled = () => false }) => (isEnabled(name) ? children : disabledElement)}
    </FeatureContext.Consumer>
  )
}

Feature.defaultProps = FeatureDefaultProps
