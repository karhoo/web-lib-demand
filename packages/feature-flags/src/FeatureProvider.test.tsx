import React from 'react'
import { shallow } from 'enzyme'
import { Feature } from './FeatureProvider'

describe('FeatureProvider', () => {
  it('should render child component if feature is enabled', () => {
    const component = shallow(
      <Feature name="test_feature">
        <div>Test</div>
      </Feature>
    )

    expect(component).toMatchSnapshot()
  })
})
