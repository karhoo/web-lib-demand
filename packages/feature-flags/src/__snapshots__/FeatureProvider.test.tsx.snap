// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureProvider should render child component if feature is enabled 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
